src/llmq/quorums_signing_shares.cpp

```diff
@@ -0,0 +1,1182 @@
+    std::map<SigShareKey, std::vector<NodeId>> nodesBySigShares;
+
+    const size_t maxRequestsForNode = 32;
+
+    // avoid requesting from same nodes all the time
+    std::vector<NodeId> shuffledNodeIds;
+    shuffledNodeIds.reserve(nodeStates.size());
+    for (auto& p : nodeStates) {
+        if (p.second.sessions.empty()) {
+            continue;
+        }
+        shuffledNodeIds.emplace_back(p.first);
+    }
+    std::random_shuffle(shuffledNodeIds.begin(), shuffledNodeIds.end(), rnd);
+
+    for (auto& nodeId : shuffledNodeIds) {
+        auto& nodeState = nodeStates[nodeId];
+
+        if (nodeState.banned) {
+            continue;
+        }
+
+        for (auto it = nodeState.requestedSigShares.begin(); it != nodeState.requestedSigShares.end();) {
+            if (now - it->second >= SIG_SHARE_REQUEST_TIMEOUT) {
+                // timeout while waiting for this one, so retry it with another node
+                LogPrint("llmq", "CSigSharesManager::%s -- timeout while waiting for %s-%d, node=%d\n", __func__,
+                         it->first.first.ToString(), it->first.second, nodeId);
+                it = nodeState.requestedSigShares.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        std::map<uint256, CSigSharesInv>* invMap = nullptr;
+
+        for (auto& p2 : nodeState.sessions) {
+            auto& signHash = p2.first;
+            auto& session = p2.second;
+
+            if (quorumSigningManager->HasRecoveredSigForSession(signHash)) {
+                continue;
+            }
+
+            for (size_t i = 0; i < session.announced.inv.size(); i++) {
+                if (!session.announced.inv[i]) {
+                    continue;
+                }
+                auto k = std::make_pair(signHash, (uint16_t) i);
+                if (sigShares.count(k)) {
+                    // we already have it
+                    session.announced.inv[i] = false;
+                    continue;
+                }
+                if (nodeState.requestedSigShares.size() >= maxRequestsForNode) {
+                    // too many pending requests for this node
+                    break;
+                }
+                bool doRequest = false;
+                auto it = sigSharesRequested.find(k);
+                if (it != sigSharesRequested.end()) {
+                    if (now - it->second.second >= SIG_SHARE_REQUEST_TIMEOUT && nodeId != it->second.first) {
+                        // other node timed out, re-request from this node
+                        LogPrint("llmq", "CSigSharesManager::%s -- other node timeout while waiting for %s-%d, re-request from=%d, node=%d\n", __func__,
+                                 it->first.first.ToString(), it->first.second, nodeId, it->second.first);
+                        doRequest = true;
+                    }
+                } else {
+                    doRequest = true;
+                }
+
+                if (doRequest) {
```

Comment:
```suggestion
                // if we got this far we should do a request
```
and make corresponding changes below